tensorboard/plugins/profile/profile_plugin.py

```diff
@@ -142,65 +144,77 @@ def index_impl(self):
     for a specific tool "x" will have a suffix name TOOLS["x"].
     Example:
       log/
-        run1/
+        model1/
           plugins/
             profile/
-              host1.trace
-              host2.trace
-        run2/
+              run1/
+                host1.trace
+              run2/
+                host2.trace
+        model2/
           plugins/
             profile/
-              host1.trace
-              host2.trace
+              run1/
+                host1.trace
+              run2/
+                host2.trace
 
     Returns:
       A map from runs to tool names e.g.
-        {"run1": ["trace_viewer"], "run2": ["trace_viewer"]} for the example.
+        {"model1,run1": ["trace_viewer"], "model1,run2": ["trace_viewer"]}
+        for the example.
     """
-    # TODO(ioeric): use the following structure and use EventMultiplexer so that
```

Comment:
So the code below seems to be doing breadth-first expansion over the entire logdir directory tree in search of profile directories, which could be extremely slow/expensive when the logdir is large and on a remote filesystem.  This has come up before in real world cases, e.g. see the description of PR #1087 (the code here is basically is doing the equivalent of tf.gfile.Walk() except by hand).

This is a concern because the code is part of index_impl() which is invoked by is_active(), and is_active() runs for every plugin - even those that aren't active - every time a user reloads their TensorBoard, so we've been pushing for them to be lightweight (see #625).

I think the best way to avoid the issue would be what this TODO suggests - using EventMultiplexer, which would mean having the profile data generation side emit an events file with a summary that serves to identify that subdirectory as a "run" with profile data (the summary doesn't need to actually contain the profile data, though doing so would be a bit more future-proof).  That way the shared EventMultiplexer logic can centralize the job of searching the directory tree for data, and the profile plugin's is_active() can just do a quick and efficient call to the EventMultiplexer to see if any profile data has been discovered yet.

Another approach that would at least alleviate the time penalty is to just push the logic here into a separate thread that runs asynchronously from is_active() as discussed on #625, but that isn't ideal because then users are still paying the cost of doing this potentially expensive traversal of the full directory hierarchy on every reload (redundantly from the main EventMultiplexer traversal) - all it fixes is the latency issue but not other possible issues like sending a lot of extra GCS API calls if the logdir is on GCS, etc.

I'm happy to discuss these options in more detail if that's helpful.

Reply:
This code simply tries to find a sub directory under root, that matches with "profile". EventMultiplexer is a good suggestion, the only thing is that we are saving traces rather than event files.  Let's discuss offline to see how we should make it work for profile plugins. Thanks.

Reply:
As discussed, let's see if we can use EventMultiplexer logic to detect the subdirectories to load data from.